src/messages.cpp

```diff
@@ -152,6 +152,86 @@ namespace quicr::messages {
         return buffer;
     }
 
+    void SerializeExtensions(Bytes& buffer,
+                             const std::optional<Extensions>& extensions,
+                             const std::optional<Extensions>& immutable_extensions)
+    {
+        Extensions combined_extensions;
+
+        // Add mutable extensions.
+        if (extensions.has_value()) {
+            for (const auto& [key, value] : *extensions) {
+                combined_extensions[key] = value;
+            }
+        }
+
+        // Serialize immutable extensions in MoQ form, and insert into combined extensions key.
+        if (immutable_extensions.has_value() && !immutable_extensions->empty()) {
+            // Immutable extensions MUST NOT contain an immutable extension entry.
+            if (immutable_extensions->contains(kImmutableExtensionsType)) {
+                throw ProtocolViolationException(
+                  "An immutable extension header must not contain another immutable extension header");
+            }
+
+            // Serialize immutable extensions.
+            Bytes immutable_data;
+            immutable_data = immutable_data << *immutable_extensions;
+            combined_extensions[kImmutableExtensionsType] = std::move(immutable_data);
+        }
+
+        // Serialize combined extensions.
+        buffer << combined_extensions;
+    }
+
+    template<class StreamBufferType>
+    bool ParseExtensions(StreamBufferType& buffer,
+                         std::optional<std::size_t>& extension_headers_length,
+                         std::optional<Extensions>& extensions,
+                         std::optional<Extensions>& immutable_extensions,
+                         std::size_t& extension_bytes_remaining,
+                         std::optional<std::uint64_t>& current_header)
+    {
+        // First, parse all extensions.
+        if (!ParseExtensions(buffer, extension_headers_length, extensions, extension_bytes_remaining, current_header)) {
+            return false;
+        }
+
+        // Extract immutable extensions if present and deserialize.
+        if (extensions.has_value() && extensions->contains(kImmutableExtensionsType)) {
+            auto immutable_data = std::move((*extensions)[kImmutableExtensionsType]);
+
+            // Remove from mutable map.
+            extensions->erase(kImmutableExtensionsType);
+
+            // Deserialize the immutable extension map.
+            auto stream_buffer = SafeStreamBuffer<uint8_t>();
```

Comment:
Doesn't look like this needs to use locking. 


```suggestion
            auto stream_buffer = StreamBuffer<uint8_t>();
```